src/ruby.rs

```diff
@@ -561,6 +560,41 @@ extern {
     pub static rb_eMathDomainError: VALUE;
 
 
+    /// Returns the length of the Ruby [`String`](rb_cString) data in bytes.
+    ///
+    /// * `string` - an instance of [`String`](rb_cString)
+    ///
+    /// # Safety
+    ///
+    /// * Undefined behavior if `string` is not a `String`
+    ///
+    /// # Ruby Documentation
+    ///
+    /// * **2.5:**
+    ///     [usage](https://ruby-doc.org/core-2.5.1/doc/extension_rdoc.html#label-Convert+VALUE+into+C+Data) |
+    ///     [spec](https://ruby-doc.org/core-2.5.1/doc/extension_rdoc.html#label-Data+Type+Conversion)
+    //+ c-macro: `#define RSTRING_LEN(str)`
+    #[link_name = "RS_RSTRING_LEN"]
+    pub fn RSTRING_LEN(string: VALUE) -> c_long;
+
+    /// Returns a pointer to the Ruby [`String`](rb_cString) data.
+    ///
+    /// * `string` - an instance of [`String`](rb_cString)
+    ///
+    /// # Safety
+    ///
+    /// * Undefined behavior if `string` is not a `String`
+    /// * The returned C string may not be nul-terminated
```

Comment:
I think we should say is _not_ nul-terminated and may contain interior nul-bytes

Reply:
From `extension.rdoc`:

> RSTRING_PTR(str) ::
>
>   String -> pointer to String data
>   Note that the result pointer may not be NUL-terminated

Reply:
I just basically copied that.

Reply:
Yeah, I think that's imprecise or even wrong, in some sense. The only correct way to interpret this pointer is _together_ with `RSTRING_LEN`, which tells you how many bytes you are allowed to look at, starting at `RSTRING_PTR`. If it happens to "end" with a nul-byte, it just means that the Ruby string contains a nul-byte at the end. (`"abc\u0000"` => RSTRING_LEN = 4, RSTRING_PTR = ['a', 'b', 'c', '\0']), which is just a special case of interior nul-byted ("ab\u0000cd").